compile.c

```diff
@@ -8915,62 +8933,116 @@ compile_op_asgn1(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node
 
         CHECK(COMPILE(ret, "NODE_OP_ASGN1 nd_rvalue: ", RNODE_OP_ASGN1(node)->nd_rvalue));
         if (!popped) {
-            ADD_INSN1(ret, node, setn, FIXNUM_INC(argc, 2+boff));
+            ADD_INSN1(ret, node, setn, INT2FIX(dup_argn+1));
         }
         if (flag & VM_CALL_ARGS_SPLAT) {
-            ADD_INSN1(ret, node, newarray, INT2FIX(1));
-            if (boff > 0) {
-                ADD_INSN1(ret, node, dupn, INT2FIX(3));
+            if (flag & VM_CALL_KW_SPLAT) {
+                ADD_INSN1(ret, node, topn, INT2FIX(2 + boff));
                 ADD_INSN(ret, node, swap);
+                ADD_INSN1(ret, node, newarray, INT2FIX(1));
+                ADD_INSN(ret, node, concatarray);
+                ADD_INSN1(ret, node, setn, INT2FIX(2 + boff));
                 ADD_INSN(ret, node, pop);
             }
-            ADD_INSN(ret, node, concatarray);
+            else {
+                ADD_INSN1(ret, node, newarray, INT2FIX(1));
+                if (boff > 0) {
+                    ADD_INSN1(ret, node, dupn, INT2FIX(3));
+                    ADD_INSN(ret, node, swap);
+                    ADD_INSN(ret, node, pop);
+                }
+                ADD_INSN(ret, node, concatarray);
+                if (boff > 0) {
+                    ADD_INSN1(ret, node, setn, INT2FIX(3));
+                    ADD_INSN(ret, node, pop);
+                    ADD_INSN(ret, node, pop);
+                }
+            }
+            ADD_SEND_R(ret, node, idASET, argc, NULL, INT2FIX(flag), keywords);
+        }
+        else if (flag & VM_CALL_KW_SPLAT) {
             if (boff > 0) {
+                ADD_INSN1(ret, node, topn, INT2FIX(2));
+                ADD_INSN(ret, node, swap);
                 ADD_INSN1(ret, node, setn, INT2FIX(3));
                 ADD_INSN(ret, node, pop);
-                ADD_INSN(ret, node, pop);
             }
-            ADD_SEND_WITH_FLAG(ret, node, idASET, argc, INT2FIX(flag));
+            ADD_INSN(ret, node, swap);
+            ADD_SEND_R(ret, node, idASET, FIXNUM_INC(argc, 1), NULL, INT2FIX(flag), keywords);
+        }
+        else if (keyword_len) {
+            ADD_INSN(ret, node, dup);
+            ADD_INSN1(ret, node, opt_reverse, INT2FIX(keyword_len+boff+2));
+            ADD_INSN1(ret, node, opt_reverse, INT2FIX(keyword_len+boff+1));
+            ADD_INSN(ret, node, pop);
```

Comment:
This rotates the stack and moves the TOS before `keyword_len+boff` from the top?
Then `dup` and `pop` seem unnecessary to me.

```suggestion
            ADD_INSN1(ret, node, opt_reverse, INT2FIX(keyword_len+boff+1));
            ADD_INSN1(ret, node, opt_reverse, INT2FIX(keyword_len+boff+0));
```